offload/plugins-nextgen/amdgpu/src/rtl.cpp

```diff
@@ -190,8 +190,9 @@ Error asyncMemCopy(bool UseMultipleSdmaEngines, void *Dst, hsa_agent_t DstAgent,
 #endif
 }
 
-Expected<std::string> getTargetTripleAndFeatures(hsa_agent_t Agent) {
-  std::string Target;
+Expected<StringRef>
+getTargetTripleAndFeatures(hsa_agent_t Agent, SmallVector<StringRef> &Targets) {
+  StringRef SpecificTarget;
```

Comment:
I'm confused here. Why do we need to care about a "specific" target vs. (potentially) non-specific ones stored in `Targets`, instead of just having them all in `Targets` and then just iterate through it?

Reply:
If you really hate to see it and make it gate landing the patch, I can remove it. As I wrote upfront, I'd like to ensure actual code execution to prioritize specific ISA over generic ones, and thus knowing which is the specific ISA is the starting point in doing so.  Also, it helps in marking XNACK enabled w/o iterating over all ISAs one more time. So, it's not like that is totally useless code. Here, I'm assuming that specific ISA is a superset of generic ISAs. Let me know if that's not the case.

Reply:
> If you really hate to see it and make it gate landing the patch, I can remove it.

I don't hate it. I just don't see how this can help. It'd be helpful if you can stack the PR (via Graphite or something similar) that depends on this to show how your proposed solution works.

> Here, I'm assuming that specific ISA is a superset of generic ISAs. Let me know if that's not the case.

The essential issue here is, it is not OpenMP runtime's call to determine if an ISA is compatible with another. For example, on a gfx1030 GPU, if the device image has `gfx10-3-generic`, but HSA returns `gfx1030` only, OpenMP runtime should still reject this ISA instead of making the decision saying, oh I know they are compatible, so I'm gonna go ahead letting it in. This is because the HSA runtime has to support generic ISA as well. If it doesn't return a generic ISA, it means it doesn't support it. With this being said, OpenMP runtime should really just, get the ISAs from HSA, and call another function to determine if the ISA is compatible, that's all.

As for handling of XNACK or other ISA features, IIUC, they are encoded in the ISA of device image. Even if for generic ISA, it is still `gfx10-3-generic:+xnack`, which makes no difference from `gfx1030:+xnack`.

Reply:
I'm talking about the case where the app contains device images for both gfx10-3-generic and gfx1036 and HSA is returning both gfx10-3-generic and gfx1036. gfx10-3-generic image will match up with gfx10-3-generic and gfx1036 image will match up with gfx1036. Functionally, choosing either images should be fine, but I'd like to ensure actual execution happens with gfx1036 image (because the programmer intentionally decided to include that version of image to be used on gfx1036 capable device). If anyone can tell me that it is already happening, great. Please point me to the areas of code for my education. Otherwise, that's where my next exploration would be --- 1) try doing that through HSA interface? 2) try doing that through PluginManager? or 3) let amdgpu RTL do that?

If this explanation does not resonate with you, I'll just give up hunting for SpecificTarget ISA from this patch --- to get the functional implementation land sooner. In terms of functional correctness, generic target feature should work fine without it. I can follow up on the rest after the functional implementation patch lands. On the side note, checking for generic via ends_with() would be incorrect then.

Reply:
Had an offline chat with @shiltian. We agreed that by controlling the image package ordering --- all specific images before any generic images, OpenMP runtime (or HSA) doesn't have to worry about whether the matched ISA is specific or not. With that, I'll gladly drop the code to identify the specific target. Will follow up with the packager side.